### ISSUE TITLE
master process begins mysteriously timing out without any crash or stack trace shown

### ISSUE THREAD

**Issue by tsortwehttam** [HaikuTeam/animator#2]
Summary: For unknown reasons, and seemingly at random, the plumbing logs start showing that the master process times out. There is no stack trace or other indication of a crash. Master's own logs indicate that it still has an open websocket connection.

Steps to reproduce: n/a - 🍺🍺🍺 bounty offered to anyone who can get this to occur _reliably_.

Screenshots, logs, etc:

The logs generally ending up looking like this:
```
[master] [alive] /Users/zack/.haiku/projects/demo/Hearts
Websocket client connection closed (commander creator)
[master] [alive] /Users/zack/.haiku/projects/demo/Hearts
Plumbing method error: Timed out waiting for client {"alias":"master"} of /Users/zack/.haiku/projects/demo/Hearts to connect Error: Timed out waiting for client {"alias":"master"} of /Users/zack/.haiku/projects/demo/Hearts to connect
    at Plumbing.awaitFolderClientWithQuery (/Users/zack/Downloads/Haiku.app/Contents/Resources/app.asar/libs/plumbing/lib/Plumbing.js:304:35)
    at Timeout._onTimeout (/Users/zack/Downloads/Haiku.app/Contents/Resources/app.asar/libs/plumbing/lib/Plumbing.js:322:23)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5)
```
Of interest in the above example, the `Websocket client connection closed (commander creator)` indicates usually that Haiku Desktop has been refreshed. Cmd+R (manual refresh) seems part of the cause but it's unclear how exactly it messes things up.

Release version, system info, etc: 2.0.75

Account, project info, etc: n/a - Seems to happen on any project/account, although it also looks like Zack's workstation seems to get the issue a lot more frequently than I do. That is possibly just because he uses the application as a user more frequently than I do.

**Comment by tsortwehttam** [comment]
closing this in favor of trello for now

**Comment by zackbrown** [comment]
I wasn't able to find this trello card--

I just wanted to add: the cmd+r repro smells like like right direction to
me. I do that fairly often, so that could help explain why I see this more
often. Maybe it has something to do with the logic around "reusing
existing..." when relaunching the same project under the same MasterProcess
session?

**Comment by tsortwehttam** [comment]
Before I move this - have you seen this in the past week or two?

**Comment by zackbrown** [comment]
I've seen a possibly related issue this week on *dev,* (maybe the same, but more verbose on dev?) which follows the behavior of "seemingly only after refresh."  It will occasionally, seemingly randomly, pop up an electron stack trace dialog saying "cannot read length of undefined," and show that ~6-8 times in a row, then the master process will be crashed (I have to restart mono after that.)  I have never seen this 8-dialog-in-a-row issue on staging/prod.

Here's the plumbing log for this case (maybe it's a corrupted bytecode file?  I've seen this while dev'ing the transform features this week, and I've done all of this inside a single Project)

```
2017-06-06T23:27:13.993Z - info: [master] File ingested: /Users/zack/.haiku/projects/Zack/Glass4/bytecode.js
[file] bytecode.js does not have a substruct
------------------------
 Error: Substruct missing in bytecode.js
    at Model._reinitializeSubstruct [as reinitializeSubstruct] (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:207:13)
    at /Users/zack/code/haiku/mono/packages/haiku-plumbing/lib/processes/MasterProcess.js:360:53
    at file.updateContents (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1125:18)
    at Model.actualizeContentState (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:130:12)
    at updateInMemoryContentState (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:92:19)
    at File.loadCodeStructures (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:118:16)
    at _extractSubstructs (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1075:16)
    at _extractSubstructs (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1195:10)
    at _parseFile (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1073:14)
    at _parseFile (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1181:15)
    at Function.loadCodeStructures (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1071:12)
    at Model.updateInMemoryContentState (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:114:19)
    at Model.updateContents (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:90:17)
    at File.findOrCreate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1118:21)
    at callAfterUpdate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:69:18)
    at Model.callAfterUpdate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:127:37)
    at callBeforeUpdate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:67:21)
    at Model.callBeforeUpdate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:122:38)
    at Model.update (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:64:19)
    at Function.findOrCreate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:198:35)
    at File.read (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1107:19)
    at _readFile (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1057:16)
    at /Users/zack/code/haiku/mono/packages/haiku-serialization/node_modules/graceful-fs/graceful-fs.js:78:16
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:445:3)
------------------------
2017-06-06T23:27:13.995Z - error:  Error: Substruct missing in bytecode.js
    at Model._reinitializeSubstruct [as reinitializeSubstruct] (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:207:13)
    at /Users/zack/code/haiku/mono/packages/haiku-plumbing/lib/processes/MasterProcess.js:360:53
    at file.updateContents (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1125:18)
    at Model.actualizeContentState (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:130:12)
    at updateInMemoryContentState (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:92:19)
    at File.loadCodeStructures (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:118:16)
    at _extractSubstructs (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1075:16)
    at _extractSubstructs (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1195:10)
    at _parseFile (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1073:14)
    at _parseFile (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1181:15)
    at Function.loadCodeStructures (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1071:12)
    at Model.updateInMemoryContentState (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:114:19)
    at Model.updateContents (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:90:17)
    at File.findOrCreate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1118:21)
    at callAfterUpdate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:69:18)
    at Model.callAfterUpdate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:127:37)
    at callBeforeUpdate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:67:21)
    at Model.callBeforeUpdate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:122:38)
    at Model.update (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:64:19)
    at Function.findOrCreate (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/ModelClassFactory.js:198:35)
    at File.read (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1107:19)
    at _readFile (/Users/zack/code/haiku/mono/packages/haiku-serialization/src/model/File.js:1057:16)
    at /Users/zack/code/haiku/mono/packages/haiku-serialization/node_modules/graceful-fs/graceful-fs.js:78:16
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:445:3)
2017-06-06T23:27:14.027Z - info: Websocket client connection closed (controllee master)
2017-06-06T23:27:14.028Z - info: master disconnected
2017-06-06T23:27:14.029Z - info: master exiting
2017-06-06T23:27:14.029Z - warn: Trying to reconnect master for folder /Users/zack/.haiku/projects/Zack/Glass4
2017-06-06T23:27:14.035Z - info: process master created @ /Users/zack/code/haiku/mono/packages/haiku-plumbing/lib/processes/MasterProcess.js
2017-06-06T23:27:14.037Z - warn: Reconnected master for folder /Users/zack/.haiku/projects/Zack/Glass4; restarting project
2017-06-06T23:27:14.037Z - warn: Restarting client master in /Users/zack/.haiku/projects/Zack/Glass4
Websocket readyState was not open so we did not send message 0.16586255769642988
2017-06-06T23:27:14.041Z - warn: Restarted client master in /Users/zack/.haiku/projects/Zack/Glass4
2017-06-06T23:27:14.042Z - info: master closing
Websocket readyState was not open so we did not send message 0.5334865231688379
2017-06-06T23:27:14.297Z - info: --> Client action applyPropertyGroupValue being sent to timeline
Websocket readyState was not open so we did not send message 0.7919653997379226
2017-06-06T23:27:14.566Z - info: --> Client action applyPropertyGroupValue being sent to creator
Websocket readyState was not open so we did not send message 0.542358293732341
Websocket readyState was not open so we did not send message 0.8701754638080395
Websocket readyState was not open so we did not send message 0.39624287415438975
Websocket readyState was not open so we did not send message 0.768609972479616
Websocket readyState was not open so we did not send message 0.8924537366269925
Websocket readyState was not open so we did not send message 0.8659174759927226
2017-06-06T23:27:16.115Z - info: Establishing websocket connection to http://0.0.0.0:1024?type=controllee&alias=master&folder=/Users/zack/.haiku/projects/Zack/Glass4
2017-06-06T23:27:16.132Z - info: Websocket client connection opened: (controllee master) {"type":"controllee","alias":"master","folder":"/Users/zack/.haiku/projects/Zack/Glass4","url":"/?type=controllee&alias=master&folder=/Users/zack/.haiku/projects/Zack/Glass4","haiku":{}}
2017-06-06T23:27:21.268Z - info: Websocket client connection closed (controllee glass)
2017-06-06T23:27:21.273Z - info: Websocket client connection closed (controllee timeline)
2017-06-06T23:27:21.400Z - info: Websocket client connection closed (commander creator)
npm closed, exiting all!
2017-06-06T23:27:21.481Z - info: Plumbing process (62350 Plumbing.js node) terminated
2017-06-06T23:27:21.483Z - info: Teardown
2017-06-06T23:27:21.483Z - info: Sending interrupt
2017-06-06T23:27:21.483Z - info: Closing server
exiting; telling children to interrupt
2017-06-06T23:27:21.488Z - info: Plumbing process (62350 Plumbing.js node) exiting
2017-06-06T23:27:21.488Z - info: Teardown
2017-06-06T23:27:21.488Z - info: Sending interrupt
2017-06-06T23:27:21.488Z - info: Closing server
2017-06-06T23:27:21.490Z - info: Process 62534 ProcessBase.js node got interrupt signal
```

**Comment by tsortwehttam** [comment]
Ok great, that's extremely useful. I'll move this into trello a bit later.

**Comment by zackbrown** [comment]
![image](https://user-images.githubusercontent.com/2100885/26885267-131395a4-4b57-11e7-819a-d1ec13ecdcbe.png)
Just repro'd on prod!

Project is Glass4 in org Zack:  zack+zack@haiku.ai | supersecure

**Comment by zackbrown** [comment]
(given the dialog, which we haven't seen before, it seems like this may be separate from the potentially ongoing phantom MasterProcess crash.  could just be corrupted bytecode?)

**Comment by zackbrown** [comment]
just ran into that electron popup again on a different project, demo/hearts.  same substruct error

![image](https://user-images.githubusercontent.com/2100885/26899036-b8382e5c-4b82-11e7-8863-d2a185735e0e.png)

**Comment by zackbrown** [comment]
(this is dev @ 2.0.109)

**Comment by tsortwehttam** [comment]
Purely out of curiosity: When you refresh, is that strictly while developing, or is there a reason you are refreshing it in production too?

**Comment by tsortwehttam** [comment]
Can you give any info you have about how to repro this?

I've tried a couple of times with the following process:

1. Start dev on mono with `npm run start -- -- --mode=blank` (straight to a blank project instead of having to go through the project loading screen)
2. Instantiate a simple element on stage
3. Move the element around, etc
4. Refresh Haiku
5. Goto 3 (do this loop about 20 times)

I can't seem to get master to crash. I'm wondering if there's any specific thing that you can remember doing that seems relevant.

**Comment by tsortwehttam** [comment]
Just note to self after exploring this a bit more - the "Cannot find property 'length'" is just a symptom of the real problem - whatever crashes master - but not really related to the underlying cause. (That said, this highlights that there are some coding defects that I will need to fix.)

Clues to the crash: `reinitializeSubstructs` expects that `ingestOne` will have populated the "substruct" (bytecode+AST data) of the bytecode file. But apparently that is not the case - there may be a race, possibly more apparent on a slower machine, in which that data gets clobbered between the operations.

**Comment by zackbrown** [comment]
bit more info from my haikuing yesterday:

- I believe this happened to me on a fresh launch of dev (no refresh)
- I was running full dev mode (npm start)
- it happened a handful of times, both while transforming items on stage
and dragging timeline segments around
- I spent maybe 45 minutes re-choreographing heartstry4, so I was working
on an actual project for a significant amount of time


also, I do sometimes refresh prod; only consistent case I can think of is
when I switch accounts (Haiku logout, cmd r, log back in)


On Jun 8, 2017 6:28 AM, "Matthew Trost" <notifications@github.com> wrote:

Just note to self after exploring this a bit more - the "Cannot find
property 'length'" is just a symptom of the real problem - whatever crashes
master - but not really related to the underlying cause. (That said, this
highlights that there are some coding defects that I will need to fix.)

Clues to the crash: reinitializeSubstructs expects that ingestOne will have
populated the "substruct" (bytecode+AST data) of the bytecode file. But
apparently that is not the case - there may be a race, possibly more
apparent on a slower machine, in which that data gets clobbered between the
operations.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/HaikuTeam/mono/issues/2#issuecomment-307103004>, or mute
the thread
